File: programs/mergesort_rec2.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

#define ARR_SIZE (25)
#define ZERO (0)
#define ONE (1)
#define TWO (2)

/*
 * Uses a better approach with a recursive switch
 * as found in wikipedia
 */

// Prints the array and checks for sortedness
int printarray(int * array) {
    int i;
    int sorted = 1;
    int prev = -1;
    for(i = 0; i < ARR_SIZE; ++i) {
	printf("%d, ", array[i]);
	if (array[i] < prev) {
	    sorted = 0;
	}
	prev = array[i];
    }
    if (sorted) {
	printf("\nok\n");
	return 0;
    } else {
	printf("\nNOT sorted\n");
	return -1;
    }

    return -1;
}


// Merges two sorted arrays into a single sorted array
void merge_(int arr1[], int size1, int arr2[], int size2, int res[]) {
    int i = 0, j = 0, k = 0;

    while (i < size1 && j < size2) {
	if (arr1[i] <= arr2[j]) {
	    res[k] = arr1[i];
	    i++;
	    k++;
	} else {
	    res[k] = arr2[j];
	    j++;
	    k++;
	}

    }
    while (i < size1) {
	res[k] = arr1[i];
	i++;
	k++;
    }
    while (j < size2) {
	res[k] = arr2[j];
	j++;
	k++;
    }
}

// Prepares the 2D array A to be fed into the actual merge_ function
// returns an integer to identify which array to find the sorted array in
void merge(int A[], int low1, int high1, int low2, int high2, int B[]) {
    merge_(&A[low1], high1-low1+1, &A[low2], high2-low2+1, &B[low1]);
}

// Mergesort in O(n) space using 2 arrays.
// returns and integer to identify which array to find the sorted array in
void mergesort_rec (int A[], int h, int low, int high, int B[]) {
    int div;

    if (low > high) {
	printf("FATAL: low: %d, high: %d\n", low, high);
        return;
    }

    if (low == high) {
        return;
    }

    div = (low + high)/2;
    mergesort_rec(B, h, low, div, A);
    mergesort_rec(B, h, div+1, high, A);

    merge(A, low, div, div+1, high, B);
}


#if 0
// Test function for merging two sorted arrays
int main() {
    int i;
    int arr[10] = {0,1,2,3,4,5,6,7,8,9};
    int res[10];
    merge_(&arr[0], 5, &arr[5], 5, res);
    for (i = 0; i < 10; i++) {
            printf("%d, ", res[i]);
    }
    printf("\n");
}

#else

// Test function to sort an array using mergesort
int main() {
    int array1[ARR_SIZE]; // indices: 0 to 49
    int array2[ARR_SIZE]; // indices: 0 to 49
    int i;

    srand(time(NULL));

    for(i = 0; i < ARR_SIZE; ++i) {
	array1[i] =
	array2[i] = rand()%100;
    }

    printarray(array1);
    mergesort_rec(array1, ZERO, 0, ARR_SIZE - 1, array2);
    return printarray(array2);

}
#endif

File: programs/mergesort_rec2.py
import random

def merge(array1, b1, len1, array2, b2, len2, final, f1):

    i = b1;
    j = b2;
    k = f1;

    while i < b1+len1 and j < b2+len2:

        if array1[i] <= array2[j]:
            final[k] = array1[i];
            i += 1;
            k += 1;
        else:
            final[k] = array2[j];
            j += 1;
            k += 1;

    while i < b1+len1:
        final[k] = array1[i];
        i += 1;
        k += 1;

    while j < b2+len2:
        final[k] = array2[j];
        j += 1;
        k += 1;


def msort(array1, b, e, array2):
    if b >= e:
        return;

    m = (b + e)/2;
    msort(array2, b,   m, array1);
    msort(array2, m+1, e, array1);

    merge(array1, b,   m-b+1,
          array1, m+1, e-(m+1)+1, array2, b);



######### main ########

print "Merge two sorted arrays" # merge two sorted arrays

arr1 = []; MAX1 = random.randint(0, 5);
arr2 = []; MAX2 = random.randint(0, 5);
final = [];

for i in range(MAX1):
    x = random.randint(1, 100);
    arr1.append(x);
    final.append(0);

for i in range(MAX2):
    y = random.randint(1, 100);
    arr2.append(y);
    final.append(0);


arr1 = sorted(arr1)
arr2 = sorted(arr2)

print arr1
print arr2

merge(arr1, 0, len(arr1), arr2, 0, len(arr2), final, 0);

print "Merging the above two sorted arrays"
print final, "/", len(final);
print "Checking sortedness"
print sorted(final);


print
print "Mergesort an array" # Mergesort an aray
array1 = []
array2 = []
MAX = random.randint(0, 10)

for i in range(MAX):
    num = random.randint(1, 100);
    array1.append(num);
    array2.append(num);

print array1
print "Sorting the above array"
msort(array1, 0, MAX-1, array2)
print array2
print "Checking sortedness"
print sorted(array2)




